From 8c41ecc98db4435dbe30dcf42d622776dd58299a Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Thu, 12 Jun 2025 14:06:19 +0100
Subject: [PATCH] tls: ressurect support for BORINGSSL_API_VERSION 19

The 19 level is the BORINGSSL_API_VERSION of the latest certified
BoringCrypto release. And users that pass-in repository override for
boringssl currently fail the build.

If one uses `build --override_repository=boringssl_fips=/usr/lib/boringssl-fips-static` to provide own build of BoringCrypto the build currently fails, as the desired function calls are not present.

This PR re-introduces partial code snippet that is present in 1.33 and earlier,
and was dropped in 1.34 development cycle. Meaning this PR is a
partial revert of https://github.com/envoyproxy/envoy/commit/e451caf9609b11418ec67f9d29b109e7a966a22c

I have also recently backported compat support for fips-20230428 to
the releases 1.33, 1.32, 1.31. See #39823, #39824, #39825. And the
same patches works on 1.30 and 1.29. Meaning those PRs are partial backports of https://github.com/envoyproxy/envoy/commit/e451caf9609b11418ec67f9d29b109e7a966a22c

Reading CMVP tea leaves, the BORINGSSL_API_VERSION of 32 will get
certification in January 2026, or earlier, meaning before 1.37
release.

Thus this is a request to keep build-time compat with fips-20230428 release
(BORINGSSL_API_VERSION 19) in the releases 1.34, 1.35 and 1.36.

This would enable all users, of all envoy version streams to upgrade
to fips-20230428 cert, and then continue to update their FedRAMP
autorisations to upgrade to BoringCrypto update stream.

Risk Level: Low
Testing: Compiled with override_repository pointing at BoringCrypto FIPS 2023042800
Fixes: https://github.com/envoyproxy/envoy/issues/39822
Signed-off-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
---
 .../tls/cert_validator/default_validator.cc   |  2 +
 source/common/tls/context_impl.cc             | 50 +++++++++++++++++++
 2 files changed, 52 insertions(+)

diff --git a/source/common/tls/cert_validator/default_validator.cc b/source/common/tls/cert_validator/default_validator.cc
index 74b7620218..6eef7c1226 100644
--- a/source/common/tls/cert_validator/default_validator.cc
+++ b/source/common/tls/cert_validator/default_validator.cc
@@ -573,11 +573,13 @@ absl::Status DefaultCertValidator::addClientValidationContext(SSL_CTX* ctx,
   // Set the verify_depth
   if (config_->maxVerifyDepth().has_value()) {
     uint32_t max_verify_depth = std::min(config_->maxVerifyDepth().value(), uint32_t{INT_MAX});
+#if BORINGSSL_API_VERSION >= 29
     // Older BoringSSLs behave like OpenSSL 1.0.x and exclude the leaf from the
     // depth but include the trust anchor. Newer BoringSSLs match OpenSSL 1.1.x
     // and later in excluding both the leaf and trust anchor. `maxVerifyDepth`
     // documents the older behavior, so adjust the value to match.
     max_verify_depth = max_verify_depth > 0 ? max_verify_depth - 1 : 0;
+#endif
     SSL_CTX_set_verify_depth(ctx, static_cast<int>(max_verify_depth));
   }
   return absl::OkStatus();
diff --git a/source/common/tls/context_impl.cc b/source/common/tls/context_impl.cc
index ff17b669f9..a0dd19f789 100644
--- a/source/common/tls/context_impl.cc
+++ b/source/common/tls/context_impl.cc
@@ -317,6 +317,7 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
   parsed_alpn_protocols_ = parseAlpnProtocols(config.alpnProtocols(), creation_status);
   SET_AND_RETURN_IF_NOT_OK(creation_status, creation_status);
 
+#if BORINGSSL_API_VERSION >= 21
   // Register stat names based on lists reported by BoringSSL.
   std::vector<const char*> list(SSL_get_all_cipher_names(nullptr, 0));
   SSL_get_all_cipher_names(list.data(), list.size());
@@ -333,6 +334,55 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
   list.resize(SSL_get_all_version_names(nullptr, 0));
   SSL_get_all_version_names(list.data(), list.size());
   stat_name_set_->rememberBuiltins(list);
+#else
+  // Use the SSL library to iterate over the configured ciphers.
+  //
+  // Note that if a negotiated cipher suite is outside of this set, we'll issue an ENVOY_BUG.
+  for (Ssl::TlsContext& tls_context : tls_contexts_) {
+    for (const SSL_CIPHER* cipher : SSL_CTX_get_ciphers(tls_context.ssl_ctx_.get())) {
+      stat_name_set_->rememberBuiltin(SSL_CIPHER_get_name(cipher));
+    }
+  }
+
+  // Add supported cipher suites from the TLS 1.3 spec:
+  // https://tools.ietf.org/html/rfc8446#appendix-B.4
+  // AES-CCM cipher suites are removed (no BoringSSL support).
+  //
+  // Note that if a negotiated cipher suite is outside of this set, we'll issue an ENVOY_BUG.
+  stat_name_set_->rememberBuiltins(
+      {"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256"});
+
+  // All supported curves. Source:
+  // https://github.com/google/boringssl/blob/3743aafdacff2f7b083615a043a37101f740fa53/ssl/ssl_key_share.cc#L302-L309
+  //
+  // Note that if a negotiated curve is outside of this set, we'll issue an ENVOY_BUG.
+  stat_name_set_->rememberBuiltins({"P-224", "P-256", "P-384", "P-521", "X25519", "CECPQ2"});
+
+  // All supported signature algorithms. Source:
+  // https://github.com/google/boringssl/blob/3743aafdacff2f7b083615a043a37101f740fa53/ssl/ssl_privkey.cc#L436-L453
+  //
+  // Note that if a negotiated algorithm is outside of this set, we'll issue an ENVOY_BUG.
+  stat_name_set_->rememberBuiltins({
+      "rsa_pkcs1_md5_sha1",
+      "rsa_pkcs1_sha1",
+      "rsa_pkcs1_sha256",
+      "rsa_pkcs1_sha384",
+      "rsa_pkcs1_sha512",
+      "ecdsa_sha1",
+      "ecdsa_secp256r1_sha256",
+      "ecdsa_secp384r1_sha384",
+      "ecdsa_secp521r1_sha512",
+      "rsa_pss_rsae_sha256",
+      "rsa_pss_rsae_sha384",
+      "rsa_pss_rsae_sha512",
+      "ed25519",
+  });
+
+  // All supported protocol versions.
+  //
+  // Note that if a negotiated version is outside of this set, we'll issue an ENVOY_BUG.
+  stat_name_set_->rememberBuiltins({"TLSv1", "TLSv1.1", "TLSv1.2", "TLSv1.3"});
+#endif
 
   // As late as possible, run the custom SSL_CTX configuration callback on each
   // SSL_CTX, if set.
-- 
2.48.1

